### ISSUE TITLE
After core's SafeMarkup changes, HTML tags are now output when using pm-enable

### ISSUE THREAD

**Issue by anavarre** [drush-ops/drush#1637]
Per https://www.drupal.org/node/2571695#comment-10384893 we should evaluate if the below issue requires Drush to chase recent core changes with SafeMarkup.

```
$ drush en comment
      <em class="placeholder">comment</em> is already enabled.                                                                      [ok]
There were no extensions that could be enabled.  
```

Berdir says:

> That's because drush is still using !placeholders and because unknown placeholders fall through to [#2575703: Remove default fall-through from PlaceholderTrait::placeholderFormat()](https://www.drupal.org/node/2575703) would remove that and you'd get an error about an invalid placeholder instead, somehow.

**Comment by jonhattan** [comment]
So we want to replace *all* usages of `!placeholder` with `@placeholder`.

**Comment by neclimdul** [comment]
It isn't clear to me yet but yes that _seems_ to be the path suggested by the CR associated with that issue.

**Comment by greg-1-anderson** [comment]
!placeholder used to mean no replacement.  @placeholder means escape, but do not add tags, whereas %placeholder means escape, and add em tags.

It seems that it would be a mistake to just change !var to @var everywhere in Drush. The big question I have is, what are the valid use cases for dt() calling through to t()?  Drush will use its own implementation of dt() when t() is not available.  Since dt() is really a completely different function than t() (the former being for cli output, the later for web browser output), shouldn't we just go ahead and make dt() use Drush's internal implementation all the time?  Are there any example code paths that would fail if we did this?

c.f. #1619. Seems that site-install passes some dt() output into the Drupal installer, the result of which is an object in the form field rather than a string. We talked about fixing that with a typecast; if the typecast is okay to put in, then perhaps avoiding Drupal's translation layer altogether would be preferable.

**Comment by neclimdul** [comment]
Yeah, its a bit confused as well honestly. 

https://www.drupal.org/node/2506427 is the main issue for removing !placeholder.

**Comment by greg-1-anderson** [comment]
Yes, so, if we keep calling Drupal's translation APIs from dt(), then in D8 we should not pass in any !var, because those are not allowed any longer.  However, there is no reason why Drush should not continue to use !var, as long as we do not call Drupal's translation API.  Really, all of Drush's placeholders should be !var, because there is never any need to cli terminal output to be html escaped.

The question is, are there any use cases at all where it is necessary to call through to the Drupal translation API, rather than using the internal Drush API?  As far as I know, this could just be an artifact left over from Drush 1.x, when Drush was a module.  I don't really see any reason for doing it in Drush 2.x or later.

**Comment by weitzman** [comment]
We reuse t() in order to reuse all the work that translators may have done on common words like 'project', 'status', etc. This gets them some translation for free. I dont know of anyone who is actually localizing drush-specific strings.

**Comment by greg-1-anderson** [comment]
I guess we have to choose between:

1) Translate !var and '!var' => $value to @var on the fly for D8.
2) Change all of our !var to @var everywhere in the code (plus contrib Drush extensions).
3) Stop translating in Drush.

Isn't it the case that t() only translates strings that are an exact match to a value that was translated? Seems that by using t(), we only get translation for strings that come from Drupal APIs, or strings that we copied exactly from Drupal source code into Drush.  There probably aren't any of the second.

My thought is that we should just keep !var, and stop translating.

**Comment by greg-1-anderson** [comment]
It occurs to me that passing through t() is probably necessary to run site-install in other languages, since the Drupal installer expects op to be passed through t().  I therefore think that we should merge the patch in #1619, and translate !var to @var everywhere it occurs in Drush core.  If the D8 translation code continues to replace !var with %var, and otherwise works, then this is probably sufficient for contrib Drush extensions.  If D8 starts rejecting !var replacements, then we should probably fix up !var with @var as a service in dt().

**Comment by neclimdul** [comment]
Makes sense. Core is looking at [removing it](https://www.drupal.org/node/2575703) but I don't think that needs to happen right away.

**Comment by greg-1-anderson** [comment]
Fixed by #1647

**Comment by jonhattan** [comment]
For the record we spoke about t() years back in drupal.org [Issue #1162694: Switch from t() to gettext()](https://www.drupal.org/node/1162694).
